File: 80-remove-duplicates-from-sorted-array-ii/remove-duplicates-from-sorted-array-ii.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        w, r = 0, 0


        while r < len(nums):
            count = 0
            curNum = nums[r]
            while r < len(nums) and nums[r] == curNum:
                r += 1
                count +=1
            
            for _ in range(min(2, count)):
                nums[w] = curNum
                w += 1

        return w

            

        

File: 80-remove-duplicates-from-sorted-array-ii/remove-duplicates-from-sorted-array-ii.dart
class Solution {

int removeDuplicates(List<int> nums) {
  if (nums.isEmpty) return 0; // Handle the edge case where the list is empty

  int writeIndex = 0; // Initialize the write index to 0

  for (int i = 0; i < nums.length; i++) {
    if (writeIndex < 2 || nums[i] != nums[writeIndex - 2]) {
      // If the write index is less than 2 or the current element is different from
      // the element two positions before the write index
      nums[writeIndex] = nums[i]; // Overwrite the write index with the current number
      writeIndex++; // Move the write index forward
    }
  }

  return writeIndex; // Return the length of the modified list
}

}

// herhangi bir sayiya 2. kez denk gelirsek w++ ve ustune yaz
// ayni sayiya 2. ken denk geldigimi bilmek iicn bir map tutacagiz.

//                    r
// [0,0,1,1,2,3,3,3,3]
//              w